Add message mapping to scalar (#61)
decentralized-identity/bbs-signature: draft-bbs-signatures.md

: points on G1 and G2 respectively. For a pairing-friendly curve, this document denotes operations in E1 and E2 in additive notation, i.e., P + Q denotes point addition and x \* P denotes scalar multiplication. Operations in GT are written in multiplicative notation, i.e., a \* b is field multiplication.

hash\_to\_curve\_g1(ostr) -> P
: The cryptographic hash function that takes an arbitrary octet string input and returns a point in G1 as defined in [@!I-D.irtf-cfrg-hash-to-curve]. The algorithm first requires selection of the pairing friendly curve and digest
algorithm, once selected apply the isogeny simplified SWU map to compute a point in G1 using the random oracle method. The domain separation tag value is dst.

hash\_to\_curve\_g2(ostr) -> P
: The cryptographic hash function that takes an arbitrary octet string input and returns a point in G2 as defined in [@!I-D.irtf-cfrg-hash-to-curve]. The algorithm first requires selection of the pairing friendly curve and digest algorithm, once selected the isogeny simplified SWU map to compute a point in G2 using the random oracle method. The domain separation tag value is dst.

point\_to\_octets\_min(P) -> ostr

> : returns the canonical representation of the point P as an octet string in compressed form. This operation is also known as serialization. (FIXME: This either requires a normative wire format or a reference to the normative wire format)

Reviewer: consider adding an issue for the FIXME, so can be tracked during issue review,

Reviewer 2: @OR13 this issue has already been addressed and updated in the latest draft see https://identity.foundation/bbs-signature/draft-bbs-signatures.html#section-1.2 PR is just out of date

8. return SK
```

## MapToScalar

MapToScalar algorithm takes an arbitrary value and maps it to a scalar value. 

> Data can be mapped multiple ways depending on how it will be used with other Zero-Knowledge Proofs. For example, it is not useful to hash an integer to a scalar using SHA256 and expect it to work properly with Range Proofs.

Reviewer 2: ```suggestion
Data to be signed can be mapped multiple ways depending on how it will be used downstream in various Zero-Knowledge Proof protocols. For example to be able to construct a range proof for a signed message, it must have been encoded in a particular way during the signing procedure.
```

> The following is a list of mappings and use cases:

Reviewer 2: Think we should talk about usecases elsewhere

```suggestion
The following is a list of mappings:
```

1. **Hashed**: for arbitrary length data fields that will not fit in the base field e.g. images, strings, biometrics, blockchain transaction info. 
   1. Use [hash\_to\_field](https://www.ietf.org/archive/id/draft-irtf-cfrg-hash-to-curve-13.html#name-hash_to_field-implementatio) with the domain separation tag `BBS_PLUS_BLS12381FQ_XOF:SHAKE-256_`.

> 2. **Bytes**: a value that is already in the base field. No mapping metho required.

Reviewer 2: ```suggestion
2. **Bytes**: a value that is already in the base field. No mapping method required.
```

3. **Numbers**: for range proofs. The value is zero-centered by take computing the base field modulus `z`=`q` / 3 integer division as the zero center and adding the positive number or substracting the negative number. To support complex numbers like decimal, the number is converted to fixed point arithmetic. The [Q format](https://en.wikipedia.org/wiki/Q_(number_format)) is used for these numbers as Q64.160. This leaves two bytes to avoid numbers greater than `q`. While it does not represent the full breadth of IEEE754 numbers, it does give a considerable resolution -2<sup>63</sup> to 2<sup>63</sup>-2<sup>-160</sup> signed and 0 to 2<sup>64</sup>-2<sup>-160</sup> unsigned or about 48 decimal places of precision. If decimals are rounded to the nearest integer, then Q format is not necessary.

> 4. **Null**: Hard coded as 5. Means the value is not included or not used.

Reviewer 2: Can you explain the assignment of the value 5?

Author: Its the smallest safe prime. Otherwise there's no other special meaning.
